server/service/client_mdm.go

```diff
@@ -30,3 +43,148 @@ func (c *Client) RequestAppleCSR(email, org string) (*fleet.AppleCSR, error) {
+	verb, path := "POST", "/api/latest/fleet/mdm/apple/bootstrap"
+
+	var b bytes.Buffer
+	w := multipart.NewWriter(&b)
+
+	// add the package field
+	fw, err := w.CreateFormFile("package", pkg.Name)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(fw, bytes.NewBuffer(pkg.Bytes)); err != nil {
+		return err
+	}
+
+	// add the team_id field
+	if err := w.WriteField("team_id", fmt.Sprint(pkg.TeamID)); err != nil {
+		return err
+	}
+
+	w.Close()
+
+	response, err := c.doContextWithBodyAndHeaders(context.Background(), verb, path, "",
+		b.Bytes(),
+		map[string]string{
+			"Content-Type":  w.FormDataContentType(),
+			"Accept":        "application/json",
+			"Authorization": fmt.Sprintf("Bearer %s", c.token),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("do multipart request: %w", err)
+	}
+
+	var bpResponse uploadBootstrapPackageResponse
+	if err := c.parseResponse(verb, path, response, &bpResponse); err != nil {
+		return fmt.Errorf("parse response: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) EnsureBootstrapPackage(bp *fleet.MDMAppleBootstrapPackage, teamID uint) error {
+	isFirstTime := false
+	oldMeta, err := c.GetBootstrapPackageMetadata(teamID)
+	if err != nil {
+		// not found is OK, it means this is our first time uploading a package
+		if !errors.Is(err, notFoundErr{}) {
+			return fmt.Errorf("getting bootstrap package metadata: %w", err)
+		}
+		isFirstTime = true
+	}
+
+	if !isFirstTime {
+		// compare checksums, if they're equal then we can skip the package upload.
+		if bytes.Equal(oldMeta.Sha256, bp.Sha256) {
+			return nil
+		}
+
+		// similar to the expected UI experience, delete the bootstrap package first
+		err = c.DeleteBootstrapPackage(teamID)
+		if err != nil {
+			return fmt.Errorf("deleting old bootstrap package: %w", err)
+		}
+	}
+
+	bp.TeamID = teamID
+	if err := c.UploadBootstrapPackage(bp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) ValidateBootstrapPackageFromURL(url string) (*fleet.MDMAppleBootstrapPackage, error) {
+	if err := c.CheckMDMEnabled(); err != nil {
+		return nil, err
+	}
+
+	return downloadRemoteMacosBootstrapPackage(url)
+}
+
+func downloadRemoteMacosBootstrapPackage(url string) (*fleet.MDMAppleBootstrapPackage, error) {
+	resp, err := http.Get(url) // nolint:gosec // we want this URL to be provided by the user. It will run on their machine.
+	if err != nil {
+		return nil, fmt.Errorf("downloading bootstrap package: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("the URL to the bootstrap_package doesn't exist. Please make this URL publicly accessible to the internet.")
+	}
+
+	// try to extract the name from a header
+	filename := "bootstrap-package.pkg"
+	cdh, ok := resp.Header["Content-Disposition"]
+	if ok && len(cdh) > 0 {
+		_, params, err := mime.ParseMediaType(cdh[0])
+		if err == nil {
+			filename = params["filename"]
+		}
```

Comment:
@georgekarrv @noahtalerman this is what we discussed re: try to extract a name from a header, and default to `bootstrap-package.pkg` if we can find anything.